File: R/TimeSeries/III - Previsao.r
# previsão com media da amostra não é legal
AirPassengers
mean(AirPassengers) # 280

# media do ano de 1960
mean(window(AirPassengers, c(1960, 1), c(1960, 12))) #476

### para fazer forecast
# install.packages('forecast')
library(forecast)

# 2 meses
# 1 no começo fica NA e 1 no final 
ma(AirPassengers, order = 2)

#  mes(es)
# se passar 1 ele calcula a media do proprio mes retornando os mesmos dados
ma(AirPassengers, order = 1)

# 12 - meses para calcular a media
# 6 no começo ficam NA e 6 no final
mediamovel = ma(AirPassengers, order = 12)

previsao = forecast(mediamovel, h = 12)

# media movelm por ser media faz suavização no grafico
plot(previsao)

###

arima = auto.arima(AirPassengers)
previsao2 = forecast(arima, h = 12)

# por se arima e não media não há suavização
plot(previsao2)

File: Python/TimeSeries/III - Previsao.py
import pandas as pd
import matplotlib.pyplot as plt

dateparse = lambda dates: pd.datetime.strptime(dates, '%Y-%m')

AirPassagers = pd.read_csv('AirPassengers.csv', sep=',',
                           parse_dates = ['Month'],
                           index_col = 'Month',
                           date_parser = dateparse)


ts = AirPassagers['#Passengers']
                  
# media da amostra
ts.mean()

# media do ano de 1960
# ts['1960'].mean()
# ts[str(ts.index.max().year)].mean()
ts['1960-01-01':'1960-12-01'].mean()

# media movel
# calcular a media com 11 meses antes + o atual
mv = ts.rolling(window = 12).mean()
ts[0:12].mean()
ts[1:13].mean() 
# ....

# traz a média do mês anterior + o atual
# ts.rolling(window = 2).mean()

# traz a media dele mesmo
# ts.rolling(window = 1).mean()

plt.plot(ts) # dados como são
plt.plot(mv, color = 'red') # media movel

###

# previsão - forecast

previsoes = []
for i in range(1, 13):
    superior = len(mv) - i
    inferior = superior - 11
    previsoes.append(mv[inferior:superior].mean())
    
previsoes = previsoes[::-1]
plt.plot(previsoes)
    

####

# Arima
from statsmodels.tsa.arima_model import ARIMA

plt.plot(ts)

# p, q, d
modelo = ARIMA(ts, order=(2, 1, 2))
modelo_treinado = modelo.fit()

modelo_treinado.summary()

# tuple
previsoes = modelo_treinado.forecast(steps = 12)

prev_prox_12_meses = previsoes[0]

eixo = ts.plot()
modelo_treinado.plot_predict(
        '1960-01-01', '1962-01-01',
        ax = eixo, plot_insample = True)


###

# arima 2???
# incompleto não instala
# aula 91
from pyramid.arima import auto_arima























    












